### PR TITLE
Introduce methods to `Field` to provide data as `CelloArray` views

### PR DESCRIPTION
**Pull request by mabruzzo**
# Overview

~~This PR depends on changes introduced in PR #113 (this specifically depends on some changes to `const` behavior of `CelloArray`s). At the time of creating this PR, all proposed changes are *entirely* isolated to the final 3 commits. Since the changes are cleanly organized among these 3 commits, this PR could probably be reviewed at the same time as PR #113.~~ EDIT: there's nothing stopping the review of this PR

This PR substantially simplifies the process of creating `CelloArray`s that wrap field data. Specifically, this PR introduces the `values_view` and `coarse_view` methods to the `FieldData` and `Field` classes. It also removes the `EnzoFieldArrayFactory` (which was used for this explicit purpose). More detail is provided below.

# `Field::values_view`

This method acts as an analogue to both the `FieldData::values` and `Field::unknowns` functions. `Field::values_view` has the following API:

```c++
template<class T>
CelloArray<T, 3> Field::values_view(int id_field, ghost_choice choice,
                                    int index_history=0) throw();

template<class T>
CelloArray<const T, 3> Field::values_view(int id_field, ghost_choice choice,
                                          int index_history=0) const throw();

template<class T>
CelloArray<T, 3> Field::values_view(std::string name, ghost_choice choice,
                                    int index_history=0) throw();

template<class T>
CelloArray<const T, 3> Field::values_view(std::string name, ghost_choice choice,
                                          int index_history=0) const throw();
```

The second argument expects a value from the following scoped enum:

```c++
enum struct ghost_choice {
  exclude, // exclude the ghost zone
  include, // include the ghost zone
  permit   // include the ghost zone, if allocated
};
```

To match the current behavior of `Field::values` and `Field::unknowns`, you would pass `ghost_choice::permit` and `ghost_choice::exclude`, respectively. However, most people using this function (to implement `Method` classes) should usually pass `ghost_choice::include` (the functions will raise an error if the ghost zones aren't allocated).

I choose not to make a separate `Field::unknowns_view` method because `Field::unknowns` is currently used in 4 places throughout the codebase, outside of the data layer (and tests). Including it would have introduced significantly more clutter to the header files.

The new api is a little verbose. To access a view of the `"density"` field, we could execute:

```c++
  CelloArray<enzo_float,3> density;
  density = field.values_view<enzo_float>("density", ghost_choice::include);
```

The analogous old-style code would look like:

```c++
  enzo_float* density;
  density = (enzo_float*)field.values("density");
```

Some optional changes we could make to reduce the function calls are offered in the following spoiler tag. I'd particularly appreciate feedback on the first 2 of them.

<details>
  <summary>Ideas for less verbose API </summary>

  - renaming the method from `values_view` to simply be called `view`

  - We could make `ghost_choice::include` the default argument of `Field::values_view`. The majority of the time, this is the behavior people expect when writing methods. As an aside, the fact that `ghost_choice` is a scoped enum (which can't be implicitly cast to an `int`), would prevent people from accidentally passing the wrong arguments. This would let us write:
  ```c++
    CelloArray<enzo_float,3> density;
    density = field.values_view<enzo_float>("density");
  ```

  - If we adopt the preceeding change, we could theoretically add an additional set of overloads that permute the order of `choice` and `index_history` arguments. But, this seems at least somewhat inadvisable since it kind of makes the number of required declarations explode explode.

  
</details>

The fact that we need to explicitly specify the field's datatype as a template argument is fairly unavoidable. This is basically analagous to the cast done with the old API, but will raise an error if there's a type mismatch. To avoid compile-time overhead associated with this method, I defined the main helper function that does most of the heavy lifting in `data_FieldData.cpp`. However, this means that when a user passes an invalid datatype to `values_view` (i.e. a non floating-point type), they might get a link-time error complaining that the function definition doesn't exist.

The last noteworthy property of this API is that the code aborts with an error message if you try to read a field that doesn't exist. This is my prefered behavior (since it means less work in the majority of cases). However, behavior more consistent with `Field::values` would be to return an empty `CelloArray<T, 3>` (you check if the array is empty with `array.is_null()`)

# `Field::coarse_view`

Our `coarse_view` methods have function signatures that much more closely follow the function signatures of the `coarse_values` methods. For example, the `Field` class defines:

```c++
template<class T>
CelloArray<T, 3> Field::coarse_view(int id_field) throw();

template<class T>
CelloArray<const T, 3> Field::coarse_view(int id_field) const throw();
```

There is no reason for this function to accept a `ghost_choice` value since `coarse_view` **ALWAYS** returns a coarse field with a single ghost zone. We following existing conventions in `Field` and `FieldData` for accepting an `index_history` argument and providing overloads that accept field names.

`coarse_view` has the same behavior as `values_view` when a name/id is specified for a field that does not exist or when the wrong datatype is specified.

# Other Related Changes

- added tests to `test_Field` to test the new API

- updated the "Accessing Field data" subsection of the developer guide on the website to reflect this new API. Let me know if you have issues with the language used to distinguish between the "traditional" and "preferred" approaches (I was mostly just using the existing language).

- I replaced all usage of the `EnzoFieldArrayFactory` class, with these new methods and have removed `EnzoFieldArrayFactory`  entirely from the codebase.

# Other misc changes:

While working on these changes I made a couple of minor improvement to some methods of `Field` and `FieldData` (these are mostly superficial changes and don't impact how the methods work):

- added an `ASSERT` check to `FieldData::coarse_values` to raise an error if `index_history!=0` (`0` is the default value), since it ignores the value of `index_history`.

- fixed an error in the type signature for `const char * Field::coarse_values (int id_field) const throw ()`

- updated docstrings for the `size` and `dimensions` methods of `Field` and `FieldData`. The updates should now make the difference between these methods more obvious. (Note: the docstring of `Field::dimensions` was previously incorrect).

# Future Work

In the future, it might make sense to more deeply integrate the `CelloArray`s more deeply into the `FieldData` class. Moreover, there are are a lot of similarities between the features of `EnzoEFltArrayMap` and the `Field` class.

### CODE REVIEW

**Comment by stefanarridge** [comment]
Not sure if I'm in a good position to review this PR since I haven't explored this part of the codebase, but based on your description here and in the documentation you added, these look like good changes to the Field API.

Regarding your ideas on making the API less verbose, I would support the first two ideas, and on the third, I would agree that adding more overloads may not be the best idea.

How exactly is `coarse_view` related to `view_values`? I ask because it seems to me that it would be good to maintain consistency between the naming of the two methods. So, either rename `view_values` to `view` as you suggested, or keep `view_values` as is, and rename `coarse_view` to `coarse_view_values`.

File: doc/source/devel/coding-guidelines.rst [hunk 1]
```diff
@@ -129,25 +131,102 @@ Summary of field attributes
...
+   int my = d.shape(1);
+   int mx = d.shape(2);
+
+   for (int iz=gz; iz<mz-gz; iz++) {
+     for (int iy=gy; iy<my-gy; iy++) {
+       for (int ix=gx; ix<my-gx; ix++) {
+         d(iz, iy, ix) = 0.0;
+       }
+     }
+   }
```
[the remark this answers is not in the document]
Comment (mabruzzo) [reply doc/source/devel/coding-guidelines.rst:229]: Good idea

**Comment by mabruzzo** [comment]
@gregbryan I'm leaning towards to make the following two changes. Do you have any strong feelings about either of them?
- make `ghost_choice::include` the default argument of `Field::values_view`. The majority of the time, this is the behavior people expect when writing methods. As an aside, the fact that `ghost_choice` is a scoped enum (which can't be implicitly cast to an `int`), would prevent people from accidentally passing the wrong arguments. This would let us write:
```c++
  CelloArray<enzo_float,3> density;
  density = field.values_view<enzo_float>("density");
```
- Renaming the method `Field::values_view` to `Field::view`

**Comment by gregbryan** [comment]
> @gregbryan I'm leaning towards to make the following two changes. Do you have any strong feelings about either of them?
> 
> * make `ghost_choice::include` the default argument of `Field::values_view`. The majority of the time, this is the behavior people expect when writing methods. As an aside, the fact that `ghost_choice` is a scoped enum (which can't be implicitly cast to an `int`), would prevent people from accidentally passing the wrong arguments. This would let us write:
> 
> ```c++
>   CelloArray<enzo_float,3> density;
>   density = field.values_view<enzo_float>("density");
> ```
> 
>  * Renaming the method `Field::values_view` to `Field::view`

I like both of these changes -- I agree that including ghost zones is what you almost always want to do and I definitely think it is useful to provide these kinds of implicit hints to developers.  

Renaming `values_view` to `view` seems sensible as well.

**Review by forrestglines — APPROVED** [review]
These look like good changes to the API. Any changes to make adoption of `CelloArrays` easier would be beneficial.

On the topic of your two questions about API simplifications, if that conversation is still relevant, I agree with both of them.

I don't see any other use of a member variable `field.view` other than this use case, and defaulting to the entire data block seems reasonable.

File: doc/source/devel/coding-guidelines.rst [hunk 2]
```diff
@@ -129,26 +131,106 @@ Summary of field attributes
...
+
+   // we can get the array shape directly from the array (a minor design quirk
+   // may make the arguments for the shape method seem a little unintuitive)
+   int mz = d.shape(0);
+   int my = d.shape(1);
+   int mx = d.shape(2);
+
+   for (int iz=gz; iz<mz-gz; iz++) {
+     for (int iy=gy; iy<my-gy; iy++) {
+       for (int ix=gx; ix<my-gx; ix++) {
```
[the remark this answers is not in the document]
Comment (mabruzzo) [reply doc/source/devel/coding-guidelines.rst:226]: The array `d` is actually unaware of it's ghost zones. That information is actually encoded in the `Field` class (internally, the `CelloArray` just tracks the stride along each axis).

But, this is definitely something we can definitely revisit in the future

**Pull request merged**